datafusion/proto/src/bytes/mod.rs

```diff
@@ -104,95 +96,106 @@ impl Serializeable for Expr {
-                _udaf: Arc<AggregateUDF>,
-            ) -> Result<Option<Arc<AggregateUDF>>> {
-                datafusion_common::internal_err!(
-                    "register_udaf called in Placeholder Registry!"
-                )
-            }
-            fn register_udf(
-                &mut self,
-                _udf: Arc<datafusion_expr::ScalarUDF>,
-            ) -> Result<Option<Arc<datafusion_expr::ScalarUDF>>> {
-                datafusion_common::internal_err!(
-                    "register_udf called in Placeholder Registry!"
-                )
-            }
-            fn register_udwf(
-                &mut self,
-                _udaf: Arc<WindowUDF>,
-            ) -> Result<Option<Arc<WindowUDF>>> {
-                datafusion_common::internal_err!(
-                    "register_udwf called in Placeholder Registry!"
-                )
+
+            fn try_decode(
+                &self,
+                _buf: &[u8],
+                _inputs: &[LogicalPlan],
+                _ctx: &TaskContext,
+            ) -> Result<Extension> {
+                unimplemented!()
             }
 
-            fn expr_planners(&self) -> Vec<Arc<dyn ExprPlanner>> {
-                vec![]
+            fn try_encode(&self, _node: &Extension, _buf: &mut Vec<u8>) -> Result<()> {
+                unimplemented!()
             }
 
-            fn udafs(&self) -> std::collections::HashSet<String> {
-                std::collections::HashSet::default()
+            fn try_decode_table_provider(
+                &self,
+                _buf: &[u8],
+                _table_ref: &TableReference,
+                _schema: SchemaRef,
+                _ctx: &TaskContext,
+            ) -> Result<Arc<dyn TableProvider>> {
+                unimplemented!()
             }
 
-            fn udwfs(&self) -> std::collections::HashSet<String> {
-                std::collections::HashSet::default()
+            fn try_encode_table_provider(
+                &self,
+                _table_ref: &TableReference,
+                _node: Arc<dyn TableProvider>,
+                _buf: &mut Vec<u8>,
+            ) -> Result<()> {
+                unimplemented!()
             }
         }
-        Expr::from_bytes_with_registry(&bytes, &PlaceHolderRegistry)?;
+
+        // Copied from from_bytes_with_ctx below but with placeholder registry instead of
+        // default.
+        {
+            let bytes: &[u8] = &bytes;
+            let protobuf = protobuf::LogicalExprNode::decode(bytes).map_err(|e| {
+                plan_datafusion_err!("Error decoding expr as protobuf: {e}")
+            })?;
+
+            let extension_codec = PlaceholderLogicalExtensionCodec {};
+            logical_plan::from_proto::parse_expr(
+                &protobuf,
+                &TaskContext::default(),
+                &extension_codec,
+            )
+            .map_err(|e| plan_datafusion_err!("Error parsing protobuf into Expr: {e}"))?;
+        }
```

Comment:
This is a bit ugly, but it was because since we're now requiring `TaskContext` instead of `dyn FunctionRegistry`, there was no easy way within `TaskContext` to just pass through arbitrary UDFs/UDAFs/UDWFs like what `PlaceHolderRegistry` was meant to do. Instead, we achieve the same behaviour via the codec. So I'm copying the `from_bytes_with_ctx` code into here to customize the codec we pass in.